x/contracts/wasmlanche/wasmlanche-test/src/lib.rs

```diff
@@ -0,0 +1,382 @@
+// Copyright (C) 2024, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+use borsh::BorshSerialize;
+use std::{
+    cell::RefCell,
+    collections::{HashMap, HashSet},
+    path::PathBuf,
+    process::Command,
+};
+use wasmlanche::{Address, ID_LEN};
+use wasmtime::{
+    Caller, Config, Engine, Extern, Instance, Linker, Memory, Module, OptLevel, Store, StoreLimits,
+    StoreLimitsBuilder, TypedFunc,
+};
+
+const WASM_TARGET: &str = "wasm32-unknown-unknown";
+const PROFILE: &str = "release";
+const ALLOC_FN_NAME: &str = "alloc";
+
+type AllocParam = u32;
+type AllocReturn = u32;
+type AllocFn = TypedFunc<AllocParam, AllocReturn>;
+pub type UserDefinedFnParam = u32;
+pub type UserDefinedFnReturn = ();
+pub type UserDefinedFn = TypedFunc<UserDefinedFnParam, UserDefinedFnReturn>;
+type StateKey = Box<[u8]>;
+type StateValue = Box<[u8]>;
+
+pub struct StoreData {
+    call_result: Option<Vec<u8>>,
+    state: HashMap<StateKey, StateValue>,
+    limiter: StoreLimits,
+}
+
+impl StoreData {
+    #[inline]
+    fn set_result(&mut self, result: Vec<u8>) {
+        self.call_result = Some(result);
+    }
+
+    #[inline]
+    fn set<Iter: IntoIterator<Item = (StateKey, StateValue)>>(&mut self, iter: Iter) {
+        self.state.extend(iter);
+    }
+
+    #[inline]
+    fn get(&self, key: &[u8]) -> Option<&StateValue> {
+        self.state.get(key)
+    }
+
+    #[inline]
+    pub fn take_result(&mut self) -> Option<Vec<u8>> {
+        self.call_result.take()
+    }
+}
+
+pub struct Builder {
+    engine: Engine,
+    module: Module,
+    linker: Linker<StoreData>,
+}
+
+impl Builder {
+    pub fn new(crate_name: &'static str) -> Self {
+        thread_local! {
```

Comment:
why is thread_local necessary here?

Reply:
`RefCell` isn't `Sync` (required for static). `static mut` is unsafe (and apparently often unsound even in the single-threaded case). That means if I want to memoize the `new` function, I need to either use `std::thread::LocalKey<RefCell>`, a `Mutex`, or a `RwLock`.  The APIs for `std::sync::{Mutex, RwLock}` are a pain because they return results that need to be unwrapped to be used.

I believe that the benchmarks that exist right now are run in a single-thread anyway so there would be no benefit to using a `Sync` type. 

Good question though!